Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,358 @@
+    simp only [AddHom.toFun_eq_coe, coe_toAddHom]
+    obtain ⟨y, rfl⟩ := lTensor.surjective Q (hgh.surjective) z
+    rw [lTensor.inverse_of_rightInverse_apply]
+    simp only [lTensor.toFun, Submodule.liftQ_apply] }
+
+/-- For a surjective `f : N →ₗ[R] P`,
+  the natural equivalence between `Q ⊗ N ⧸ (image of ker f)` to `Q ⊗ P` -/
+noncomputable def lTensor.equiv :
+    ((Q ⊗[R] N) ⧸ (LinearMap.range (lTensor Q f))) ≃ₗ[R] (Q ⊗[R] P) :=
+  lTensor.linearEquiv_of_rightInverse  Q hfg (Function.rightInverse_surjInv hg)
+
+/-- Tensoring an exact pair on the left gives an exact pair -/
+theorem lTensor_exact : Exact (lTensor Q f) (lTensor Q g) := by
+  rw [exact_iff]
+  rw [← Submodule.ker_mkQ (p := range (lTensor Q f))]
+  rw [← lTensor.inverse_comp_rTensor Q hfg hg]
+  apply symm
+  apply LinearMap.ker_comp_of_ker_eq_bot
+  rw [LinearMap.ker_eq_bot]
+  exact (lTensor.equiv Q hfg hg).symm.injective
+
+/-- Right-exactness of tensor product -/
+lemma lTensor_mkQ (N : Submodule R M) :
+    ker (lTensor Q (N.mkQ)) = range (lTensor Q N.subtype) := by
+  rw [← exact_iff]
+  exact lTensor_exact Q (LinearMap.exact_subtype_mkQ N) (Submodule.mkQ_surjective N)
+
+
+/-- The direct map in `rTensor.equiv` -/
+def rTensor.toFun :
+  N ⊗[R] Q ⧸ range (rTensor Q f) →ₗ[R] P ⊗[R] Q := by
+  apply Submodule.liftQ _ (rTensor Q g)
+  rw [range_le_iff_comap, ← ker_comp, ← rTensor_comp,
+    hfg.linearMap_comp_eq_zero, rTensor_zero, ker_zero]
+
+/-- The inverse map in `rTensor.equiv_of_rightInverse` (computably, given a right inverse) -/
+def rTensor.inverse_of_rightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+    P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ LinearMap.range (rTensor Q f) := by
+  rw [exact_iff] at hfg
+  refine
+    TensorProduct.lift
+      { toFun := fun p ↦ Submodule.mkQ _ ∘ₗ TensorProduct.mk R _ _ (h p)
+        map_add' := fun p p' => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_
+        map_smul' := fun r p => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_ }
+  · change h (p + p') ⊗ₜ[R] q - (h p ⊗ₜ[R] q + h p' ⊗ₜ[R] q) ∈ range (rTensor Q f)
+    rw [← TensorProduct.add_tmul, ← TensorProduct.sub_tmul]
+    apply le_comap_range_rTensor f
+    simp only [← hfg, mem_ker, map_sub, map_add, hgh _, sub_self]
+  · change h (r • p) ⊗ₜ[R] q - r • h p ⊗ₜ[R] q ∈ range (rTensor Q f)
+    rw [TensorProduct.smul_tmul', ← TensorProduct.sub_tmul]
+    apply le_comap_range_rTensor f
+    simp only [← hfg, mem_ker, map_sub, map_smul, hgh _, sub_self]
+
+lemma rTensor.inverse_of_rightInverse_apply
+    {h : P → N} (hgh : Function.RightInverse h g) (y : N ⊗[R] Q) :
+    (rTensor.inverse_of_rightInverse Q hfg hgh) ((rTensor Q g) y) =
+      Submodule.Quotient.mk (p := LinearMap.range (rTensor Q f)) y := by
+  simp only [← LinearMap.comp_apply, ← Submodule.mkQ_apply]
+  rw [exact_iff] at hfg
+  apply LinearMap.congr_fun
+  apply TensorProduct.ext'
+  intro n q
+  simp only [coe_comp, Function.comp_apply, rTensor_tmul, Submodule.mkQ_apply]
```

Comment:
You can use `says` as above.

Reply:
done, but I don't know what the `says` tactic does.